runtime/src/main/java/org/corfudb/runtime/collections/CorfuTable.java

```diff
@@ -198,45 +199,60 @@ public int hashCode() {
      * @param <V>                   Value type
      * @return                      A type token to pass to the builder.
      */
-    static <K, V>
-    TypeToken<CorfuTable<K, V>>
-        getTableType() {
-            return new TypeToken<CorfuTable<K, V>>() {};
+    static <K, V> TypeToken<CorfuTable<K, V>> getTableType() {
+        return new TypeToken<CorfuTable<K, V>>() {};
     }
 
     // The "main" map which contains the primary key-value mappings.
     private final StreamingMap<K,V> mainMap;
-    private final Set<Index<K, V, ? extends Comparable>> indexSpec = new HashSet<>();
-    private final Map<String, Map<Comparable, Map<K, V>>> secondaryIndexes = new HashMap<>();
-
+    private final Set<Index<K, V, ? extends Comparable>> indexSpec;
+    private final Map<String, Map<Comparable, Map<K, V>>> secondaryIndexes;
+    private final CorfuTable<K, V> optimisticTable;
+
+    public CorfuTable(StreamingMap<K,V> mainMap,
+                      Set<Index<K, V, ? extends Comparable>> indexSpec,
+                      Map<String, Map<Comparable, Map<K, V>>> secondaryIndexe,
+                      CorfuTable<K, V> optimisticTable) {
+        this.mainMap = mainMap;
+        this.indexSpec = indexSpec;
+        this.secondaryIndexes = secondaryIndexe;
+        this.optimisticTable = optimisticTable;
+    }
     /**
      * Generate a table with a given implementation for the {@link StreamingMap}.
      */
     public CorfuTable(IndexRegistry<K, V> indices,
                       Supplier<StreamingMap<K, V>> streamingMapSupplier) {
+        this.indexSpec = new HashSet<>();
+        this.secondaryIndexes = new HashMap<>();
+        this.mainMap = streamingMapSupplier.get();
+
+        this.optimisticTable = new CorfuTable<>(this.mainMap.getOptimisticMap(), this.indexSpec,
```

Comment:
optimisitcTable breaks the abstraction of SMR. All optimstic updates assume that the initial table is empty. This is correct for CorfuTable because different keys are orthogonal, but it is not true for general SMR.

Reply:
Moreover, one single optimisticTable prevent us developing nested transactions.

Reply:
Good catch! The implementation is incomplete -- there is a JIRA ticket that references this issue. Long story short, the TransactionalContext and the OptimisticContext need to be reconciled. This is going to be a non-trivial change, so I opted not to tackle this issue in this PR.